solr/core/src/test/org/apache/solr/handler/admin/api/DeleteReplicaPropertyAPITest.java

```diff
@@ -0,0 +1,116 @@
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import io.opentracing.noop.NoopSpan;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.solr.SolrTestCaseJ4;
+import org.apache.solr.cloud.OverseerSolrResponse;
+import org.apache.solr.cloud.api.collections.DistributedCollectionConfigSetCommandRunner;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.cloud.ZkNodeProps;
+import org.apache.solr.common.util.NamedList;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+
+/** Unit tests for {@link DeleteReplicaPropertyAPI} */
+public class DeleteReplicaPropertyAPITest extends SolrTestCaseJ4 {
+
+  private CoreContainer mockCoreContainer;
+  private DistributedCollectionConfigSetCommandRunner mockCommandRunner;
+  private SolrQueryRequest mockQueryRequest;
+  private SolrQueryResponse queryResponse;
+  private ArgumentCaptor<ZkNodeProps> messageCapturer;
+
+  private DeleteReplicaPropertyAPI deleteReplicaPropApi;
+
+  @BeforeClass
+  public static void ensureWorkingMockito() {
+    assumeWorkingMockito();
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    super.setUp();
+
+    mockCoreContainer = mock(CoreContainer.class);
+    mockCommandRunner = mock(DistributedCollectionConfigSetCommandRunner.class);
+    when(mockCoreContainer.getDistributedCollectionCommandRunner())
+        .thenReturn(Optional.of(mockCommandRunner));
+    when(mockCommandRunner.runCollectionCommand(any(), any(), anyLong()))
+        .thenReturn(new OverseerSolrResponse(new NamedList<>()));
+    mockQueryRequest = mock(SolrQueryRequest.class);
+    when(mockQueryRequest.getSpan()).thenReturn(NoopSpan.INSTANCE);
+    queryResponse = new SolrQueryResponse();
+    messageCapturer = ArgumentCaptor.forClass(ZkNodeProps.class);
+
+    deleteReplicaPropApi =
+        new DeleteReplicaPropertyAPI(mockCoreContainer, mockQueryRequest, queryResponse);
+  }
+
+  @Test
+  public void testReportsErrorWhenCalledInStandaloneMode() {
+    when(mockCoreContainer.isZooKeeperAware()).thenReturn(false);
+
+    final SolrException e =
+        expectThrows(
+            SolrException.class,
+            () -> {
+              deleteReplicaPropApi.deleteReplicaProperty(
+                  "someColl", "someShard", "someReplica", "somePropName");
+            });
+    assertEquals(400, e.code());
+    assertTrue(
+        "Exception message differed from expected: " + e.getMessage(),
+        e.getMessage().contains("not running in SolrCloud mode"));
+  }
+
+  @Test
+  public void testCreatesValidOverseerMessage() throws Exception {
```

Comment:
Maybe this is too internal?  I like to think we don't need the Overseer for this -- remember distributed state updates.  FWIW I prefer that we test functionally and not internal details subject to change that make internal refactoring harder.

Reply:
For example, set the property and then get the property to see that we set it.  Don't test how it works.

Reply:
Tbh I'm at a bit of a loss as to how to proceed in testing this stuff.

I want to address your concerns as best as I can, but I feel pretty strongly that we shouldn't be writing even more integration tests every time we modify a v2 API.  And if I'm not misunderstanding you, that kindof sounds like what you're suggesting here.  (i.e. "set the property and then get the property" requires (at a minimum) a ZK for the props to live in, if not a full Solr to submit the API to)

**Why am I anti-integration test for these PRs?**

Firstly, because it just wouldn't scale.  If I add an integration test for each v2 API as I refactor it, we'll be near doubling the size of our test suite!  (Or at least, something on that order of magnitude *general handwavy-ness*.) And then, down the road when individual SolrJ classes switch over to using v2, we'll have essentially duplicate tests everywhere!

Secondly, because an integration test wouldn't add all that much coverage for the trouble - these refactor PRs change the v1 codepath to _use_ the v2 objects, so the existing v1 integration tests are already testing the v2 code, in a sense.  There's a bit of a gap around validating that the v2 endpoint itself looks the way we want it to.  A v2-specific integration test _would_ cover that, but there are lighter, quicker, less flaky ways to close that gap (e.g. a `JerseyTest`-based unit test).

Lastly, I don't see messages as internal at all.  In fact, I see them as external, almost by definition.  They're sent/received/understood by multiple sub-systems, which makes them internal to none, right?  Unless you consider even these interfaces between sub-components to be "internal"?

Ultimately if you have this thing that one part of your system throws over the wall to another, it seems helpful to have some low-level automated validation that you don't accidentally steer from that contract.  An integration test might catch that break, but it might not.  That's one of the advantages of unit tests here - they can quickly cover more variations than is usually feasible to integration test.

**So, What Now?**

I'm really hoping there's common ground to address both our concerns.  Lmk if you see some.  I mentioned some potential alternatives in the second half of my comment [here](https://github.com/apache/solr/pull/1053#discussion_r1006192774), so hopefully one of those will work?

I do really want to get this sorted out; not for the sake of this PR specifically but because this is just 1 of many similar PRs we'll need to get v2 into line, and I want to find something we can both live with in all those going forward.

I'm going to bring all the non-test stuff in this PR up to date, so that when we find a compromise on the test stuff, we'll be ready to go.

Reply:
You're right that a SolrCloud API integration test would involve running Jetty & ZK.  That is certainly a subset of the API surface area though.  For those tests, I would hope that variations of a single API (e.g. tests all its input variations) could be in the same test class using a common running instance.  For non-SolrCloud tests, especially after a great test refactor to be based on SolrClients, they would run with EmbeddedSolrServer (not Jetty) but could be toggled to a Jetty mode on a nightly job, say; even SolrCloud.  To be clear, I'm talking about one test class that is able to run in different modes based on say a system property toggle.  I do this at work now and I've done it in the past to great effect.  Maybe that's only partly related to our discussion above (how to test the API); but I want to get your reaction any way.  Of course we have real unit tests for some low-level things too :-)

If V1 is merely going to be calling V2 (yay!!), then I agree we can have real unit tests that only test that mechanism in isolation.

Still, there's how you test the primary API (V2).  We already have tests for these in general, albeit using V1?  They can be changed to use V2.

I may be talking past you in a sense... but I wanted to get this out there and maybe circle back to your ideas.

Reply:
> Maybe that's only partly related to our discussion above (how to test the API); but I want to get your reaction any way.

It seems like a reasonable way to handle things.  My main concern would be that the base class abstracts things well enough so that test code doesn't need to know/care how Solr is actually run.  That sounds hard - my off-the-top-of-the-head impression is that we have lots of tests that fetch the JettySolrRunner for some reason or other.  But if you're already doing something similar in your work context, and you feel confident in the approach, it seems solid to me. 

> If V1 is merely going to be calling V2 (yay!!), then I agree we can have real unit tests that only test that mechanism in isolation.

Yep, that's exactly what this PR does!  Of course, there's some plumbing/shim logic to convert the v1 params into the format v2 expects, etc.  But with this PR, the heart of the v1 delete-replica-prop logic is "just" a [call to DeleteReplicaPropertyAPI.deleteReplicaProperty](https://github.com/apache/solr/pull/1053/files#diff-582348d44491dcb0ce1dfb169fb544e9e95620b2d0448eb1a1744f4e8dd5a349R1313).

This is all, interestingly enough, inspired by a comment you yourself made [here](https://lists.apache.org/thread/43xy1drq8wswg6vpo9b11rkvdqtqomlv).  So, thanks haha!

> Still, there's how you test the primary API (V2). We already have tests for these in general, albeit using V1? They can be changed to use V2.

They can be switched over to v2, and that is my plan, but I don't think we're there just yet.  I really want people using v2, but v1 is what people are on today, so I think it makes sense to keep the coverage there, at least up until we near v1 deprecation.  (And of course, the whole v1-as-a-shim-on-top plan allows v1 integration tests to cover the v2 codepath implicitly, so it's not like that's untested while we defer the switch).

(I'm close to publishing a SIP that aims to paint a picture of what the path to v1-deprecation looks like.  I'm sure it's awhile away, but maybe waiting will be more bearable when we have a concrete path to the destination? idk)

Reply:
> Yep, that's exactly what this PR does! 

Can you substantiate that please?  I'm talking about a unit test that only tests the V1 to V2 mapping without implementation details of how V2 does its job.  I could imagine mocking (yes me, endorsing mocking) to capture what parameters are passed to DeleteReplicaPropertyAPI.deleteReplicaProperty() to ensure they align with a V1 call.  Mention of the Overseer would be an unwanted implementation detail.

Reply:
> a unit test that only tests the V1 to V2 mapping without implementation details of how V2 does its job

We can absolutely do that, and if you want me to I can.  But I still feel like I'm missing some nuance here.  Isn't CollectionHandler's use of DeleteReplicaPropertyAPI under-the-hood just as much of an "implementation detail" as DeleteReplicaPropertyAPI's use of the overseer?  Why is one OK to test but not the other?  If anything I'd consider the overseer as less of an internal detail as it's doing message passing to some other system!

Ideally I'd write both of those tests: one that validates CollectionHandler's usage of DRPA, and one that validates DRPA's message-crafting.

Reply:
> We can absolutely do that, and if you want me to I can. But I still feel like I'm missing some nuance here. Isn't CollectionHandler's use of DeleteReplicaPropertyAPI under-the-hood just as much of an "implementation detail" as DeleteReplicaPropertyAPI's use of the overseer?

I very much disagree.  The *API classes are *the* primary internal API going forward.  We want to test that some old V1 API thing will call it.  Whatever the *API class does internally is an implementation detail.  Today it talks to the Overseer, tomorrow, lets remove the Overseer :-). 

I'd rather no test test Overseer message crafting because that cements the Overseer into Solr further.

Reply:
OK, I don't super agree but I feel like I get the distinction.  I appreciate you taking the time to explain where you're coming from on this.  I know it's a lot of talk for a small thing, but I'm hoping being thorough here will save time on reviews going forward.

I'll try this as a compromise:

- I still feel like the message-construction code is important to unit test for now, because it's where the bulk of the "API" logic is as the code stands today.  But it doesn't have to mention the overseer or even live in the API class itself. 
 I'll refactor the message-creation code out, and that'll let us unit test it without mocks **and** it'll separate the tests from the API class itself so that they can just be nuked if we move away from the overseer (rather than playing some "cementing" role).
- We can refactor the v1 code that calls into v2 in such a way that it's testable (maybe using mocks, but hopefully not?) to ensure that the v1 codepath uses the v2 API class properly.

Is that something you could live with? @dsmiley

Reply:
That's a reasonable compromise; thanks.